### DISCUSSION TITLE
TLS overhead?

### DISCUSSION THREAD

**Discussion by binaryfire** [valkey-io/valkey#1289] [Q&A]
Hi all

I'm setting up a mesh where thousands of devices need to communicate with a Valkey instance over the internet. I'm tossing up between a Wireguard mesh or using Valkey with TLS. However I just found this issue showing ~37% performance hit with Redis when using TLS: https://github.com/redis/redis/issues/7595

Is the TLS performance hit with Valkey similar?

Thanks

**Comment by amirreza-sf80** [comment]
Hi
valkey is a fork of redis, so old problems with redis come here as well
looking at  release notes, i don't see any note about fixing this issue

about wireguard, well you should benchmark and see how much performance hit there will be

also [this](https://github.com/Snapchat/KeyDB/issues/728#issuecomment-1780388781) might be a good note

**Comment by zuiderkwast** [comment]
TLS has some overhead yes. Good news though is that with Valkey 8.0, it's possible to use TLS with IO-threading, which offloads IO and the TLS work to threads, giving more CPU for the main thread for actual command execution. Besides, IO threading has been much improved compared to the older implementation.

**Reply by amirreza-sf80** [reply]
hi
thank you! 
i think that should be reflected more on the docs/change log, since it's a fix to a known problem

**Reply by zuiderkwast** [reply]
The IO threading docs are very limited, unfortunately. We need some help with that. (@ranshid FYI.)
